app/assets/javascripts/qpixel_api.js

```diff
@@ -258,11 +249,45 @@ window.QPixel = {
       console.error(resp);
     }
     else {
-      this._preferences = data.preferences;
-      localStorage['qpixel.user_preferences'] = JSON.stringify(this._preferences);
+      await QPixel._updatePreferencesLocally(data.preferences);
     }
   },
 
+  /**
+   * Get the key to use for storing user preferences in localStorage, to avoid conflating users
+   * @returns {Promise<string>} the localStorage key
+   */
+  _preferencesLocalStorageKey: async () => {
+    const user = await QPixel.user();
+    return `qpixel.user_${user.id}_preferences`;
+  },
```

Comment:
If I have understood correctly, the localStorage version of `_preferences` is intended to always match the server.

The `setPreference` function is the only way to change a preference, and it always updates the server first, and then localStorage. So fetching the preferences from the server should always give the most up to date preferences.

In cases where two or more users share a browser, they may historically have overridden each other's localStorage preferences as this only existed once per browser. In these cases their preferences will now revert to their correct preferences from the server. If a user has become accustomed to incorrect preferences, this will appear to them as a change. Is this something we need to avoid / warn about? I don't know how many users (if any) share a browser with another user.

Reply:
Since this is my first code change to QPixel, if there's any doubt about the consequences it might be worth a second (third?) opinion. My understanding may be incorrect.

Reply:
@trichoplax your understanding sounds correct to me. The only other effect that I can think this will have is to force a re-fetch of every user's preferences the first time they visit after this is deployed. That's not a major load.